### ISSUE TITLE
ulmo module not importing cdec module correctly

### ISSUE THREAD

**Issue by hectcastro** [WikiWatershed/model-my-watershed#2491]
This exception is currently being thrown when Big CZ routes are being evaluated for health checks:

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 108, in get_response
    response = middleware_method(request)
  File "/opt/app/mmw/middleware.py", line 25, in process_request
    callback_kwargs = resolver.resolve(request.path_info)
  File "/usr/local/lib/python2.7/dist-packages/django/core/urlresolvers.py", line 365, in resolve
    for pattern in self.url_patterns:
  File "/usr/local/lib/python2.7/dist-packages/django/core/urlresolvers.py", line 401, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/usr/local/lib/python2.7/dist-packages/django/core/urlresolvers.py", line 395, in urlconf_module
    self._urlconf_module = import_module(self.urlconf_name)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/opt/app/mmw/urls.py", line 12, in <module>
    import apps.bigcz.urls
  File "/opt/app/apps/bigcz/urls.py", line 8, in <module>
    from apps.bigcz import views
  File "/opt/app/apps/bigcz/views.py", line 15, in <module>
    from apps.bigcz.clients import CATALOGS
  File "/opt/app/apps/bigcz/clients/__init__.py", line 6, in <module>
    from apps.bigcz.clients import cinergi, hydroshare, cuahsi
  File "/opt/app/apps/bigcz/clients/cuahsi/__init__.py", line 11, in <module>
    from apps.bigcz.clients.cuahsi.details import details, values  # NOQA
  File "/opt/app/apps/bigcz/clients/cuahsi/details.py", line 10, in <module>
    from ulmo.cuahsi import wof
  File "/usr/local/lib/python2.7/dist-packages/ulmo/__init__.py", line 12, in <module>
    from . import cdec
ImportError: cannot import name cdec
```

**Comment by mmcfarland** [comment]
FWIW, I can execute the import specified in the stack trace on the app VM, but I have not destroyed and reprovisioned from scratch in some time.  I know that we recently switched to a fork of ulmo and perhaps this dependency is not specified correctly, but was left in the site-packages of previous installs on dev?

```python
>>> from ulmo.cuahsi import wof
/usr/local/lib/python2.7/dist-packages/ulmo/twc/kbdi/core.py:20: FutureWarning: pandas.tslib is deprecated and will be removed in a future version.
You can access Timestamp as pandas.Timestamp
  CSV_SWITCHOVER = pandas.tslib.Timestamp('2016-10-01')
>>> wof
<module 'ulmo.cuahsi.wof' from '/usr/local/lib/python2.7/dist-packages/ulmo/cuahsi/wof/__init__.pyc'>

```

**Comment by rajadain** [comment]
The only difference between the Ulmo main and fork is this line:

https://github.com/ulmo-dev/ulmo/compare/master...emiliom:wml_values_md

which does not change any of the imports. I'm going to destroy and reprovision the App VM to see what's up, else it could just be a broken AMI

**Comment by rajadain** [comment]
Destroyed and re-tested `release/1.21.0` on my local, its working perfectly. I think we should try percussive maintenance.

**Comment by hectcastro** [comment]
> I think we should try percussive maintenance.

What does this mean?

**Comment by rajadain** [comment]
Sorry. Trigger a new AMI build and deploy it.

**Comment by rajadain** [comment]
Should this card be closed now?

**Comment by rajadain** [comment]
Closing this issue as a fresh instance seems to have fixed it. Will re-open if seen again.

**Comment by rajadain** [comment]
Opening this back up since it we saw it on production.

**Comment by rajadain** [comment]
Googling around for the error, I see that in many cases the cause has been a version conflict (e.g. see the answers here: https://stackoverflow.com/questions/23620608/django-gunicorn-importerror-cannot-import-name-get-path-info). Looking at recent things that were changed in our dependencies: https://github.com/WikiWatershed/model-my-watershed/blame/develop/src/mmw/requirements/base.txt, I see that `django-cors-headers` was also recently added. While the added version does claim to support Django 1.8 (and clearly does in most cases), could it be that is somehow part of this? Their most recent "release" on GitHub is [1.2.0](https://github.com/ottoyiu/django-cors-headers/releases/tag/1.2.0), quite an earlier version of the `2.1.0` we're using, but I thought that they had just switched to PyPI for listing releases (because 2.1.0 is the version that gets installed if you don't specify a version).

**Comment by hectcastro** [comment]
I did not pursue this (the message above this one) when trying to troubleshoot things, but my immediate instincts on it say that if this was the issue, wouldn't it be happening consistently vs. sporadically?

To log what I did try:

I was not able to reproduce this issue locally despite trying a bunch of things (Gunicorn configuration changes, import tests in isolation w/ high concurrency, local VM reboots). The only reliable way to reproduce it seems to be on EC2, with freshly booted instances of the application server AMI. Restarts of the application service seem to stabilize things, but there were some instances where it occurred again after a period of stability.

Thus far, I've had consistent success with https://github.com/WikiWatershed/model-my-watershed/commit/1d001d84d6abc70b2b3d953927c5f33674aa74b8, which takes a scoped import approach. This helps avoid evaluation of the import when URLs are being parsed/resolved (which may be occurring on all requests).

A few additional thoughts:

- The fix above does not address the root cause, but it does limit it quite a bit. It seems like a user really needs to try to execute the code where this import lives, which should minimize its impact.
- I still feel as though this is being caused by some sort of circular dependency + race condition, but haven't been able to pinpoint it.
